lib/libwireguard.py

```diff
@@ -0,0 +1,140 @@
+            os.remove(self.public_key_file)
+        if os.path.isfile(self.private_key_file):
+            private_key_contents = self.read_file(self.private_key_file)
+            self.kv.set("private-key", private_key_contents)
+            os.remove(self.private_eky_file)
+        log("Successfully migrated keys to key-value store", level="info")
+        return True
+
+    def run_wg(self, args, stdin=b""):
+        """Run wg with the supplied stdin and command, and return stdout."""
+        cmd = ["wg"]
+        if args:
+            cmd.extend(args)
+        process = Popen(cmd, stdin=PIPE, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate(stdin)
+        hookenv.log(
+            "Run wg command {}.\nstdout: {}\nstderr: {}".format(args, stdout, stderr),
+            level="debug",
+        )
+        return stdout.decode("utf-8").rstrip()
+
+    def configure_keys(self):
+        """Generate public and private keys."""
+        self.migrate_keys()
+        if self.charm_config.get("private-key"):
+            # a key has been provided, store it
+            private_key = self.charm_config.get("private-key")
+            if self.kv.get("private-key") != private_key:
+                self.kv.set("private-key", self.charm_config.get("private-key"))
+        else:
+            # key has not been provided, check if we have already generated one,
+            # if not, generate one
+            if self.kv.get("private-key"):
+                log("Private key already configured", level="debug")
+            else:
+                log("Generating private key", level="debug")
+                private_key = self.run_wg(["genkey"])
+                self.kv.set("private-key", private_key)
+
+        if not self.kv.get("public-key"):
+            # Generate a public key for the configured/generated private key
+            public_key = self.run_wg(["pubkey"], private_key.encode())
+            self.kv.set("public-key", public_key)
+
+    def configure(self):
+        """Write configuration for WireGuard."""
+        self.configure_keys()
+        self.configure_forwarding()
+
+        host.service('stop', self.service_name)
+        peers_string = base64.b64decode(self.charm_config["peers"])
+        peers_yaml = yaml.load(peers_string)
+        context = {
+            "private_key": self.kv.get("private_key"),
+            "listen_port": self.charm_config["listen-port"],
+            "peers": peers_yaml,
+            "address": self.charm_config["address"],
+            "forward": self.charm_config["forward-ip"],
+            "forward_dev": self.charm_config["forward-dev"],
+        }
+        templating.render("wg0.conf.j2", self.cfg_file, context, perms=0o660)
+        host.service("enable", self.service_name)
+        host.service("start", self.service_name)
+
+        self.configure_ports()
+        hookenv.open_port(self.charm_config["listen-port"], protocol="UDP")
```

Comment:
configure_ports is automatically calling open_port no need to manually call it here as well.